### PR TITLE
Support pattern matching to filter artifacts & merge to same directory

### PR DESCRIPTION
**Pull request by robherley**
As mentioned in the following comment:
- https://github.com/actions/upload-artifact/issues/472#issuecomment-1861571655

We're adding some additional inputs to better support use cases outlined in that issue.

### CODE REVIEW

**Comment by TWiStErRob** [comment]
Fixes #6
Related #68

**Review by jtamsut — DISMISSED** [review]

**Review by yacaovsnc — APPROVED** [review]

**Review by bethanyj28 — APPROVED** [review]

**Review by vmjoseph — APPROVED** [review]

**Review by SrRyan — APPROVED** [review]

File: src/download-artifact.ts [hunk 1]
```diff
@@ -80,23 +83,36 @@ async function run(): Promise<void> {
...
       ...options
     })
+    artifacts = listArtifactResponse.artifacts
 
-    if (listArtifactResponse.artifacts.length === 0) {
-      throw new Error(
-        `No artifacts found for run '${inputs.runID}' in '${inputs.repository}'`
+    core.debug(`Found ${artifacts.length} artifacts in run`)
+
+    if (inputs.pattern) {
```
Comment (nulano) [inline src/download-artifact.ts:90]: The else block of `if (isSingleArtifactDownload)` above,
https://github.com/actions/download-artifact/blob/f44cd7b40bfd40b6aa1cc1b9b5b7bf03d3c67110/src/download-artifact.ts#L77-L80
should probably be moved as an else block of the `if` on this line.

Currently it prints a slightly confusing message when specifying a `pattern` input: https://github.com/nulano/Pillow/actions/runs/7281978318/job/19844364694#step:3:36

**Comment by mathbunnyru** [comment]
> * Fixes [Feature request: wildcard support (or download-all) #6](https://github.com/actions/download-artifact/issues/6)
> * Fixes [download artifact matching pattern #103](https://github.com/actions/download-artifact/issues/103)
> * Fixes [[feat req] support name filter #165](https://github.com/actions/download-artifact/issues/165)
> * Fixes [[feat req] glob based name for multiple artifacts #214](https://github.com/actions/download-artifact/issues/214)
> * Related [Put all artifacts back. #68](https://github.com/actions/download-artifact/issues/68)

All these issues can be closed now 🎉 
Unfortunately, this message wasn't copied to PR description before merge, so they were not closed automatically.
Maybe someone who has permissions can close these issues?

**Comment by mathbunnyru** [comment]
@robherley thank you for this PR, this makes a huge difference for many workflows.

**Comment by TWiStErRob** [comment]
Can one of @robherley @yacaovsnc @bethanyj28 @vmjoseph @SrRyan, please action https://github.com/actions/download-artifact/pull/259#issuecomment-1861691243

**Pull request merged**